go/parquet/file/column_writer.go

```diff
@@ -0,0 +1,567 @@
+	defer values.Release()
+
+	isV1DataPage := w.props.DataPageVersion() == parquet.DataPageV1
+	if w.descr.MaxDefinitionLevel() > 0 {
+		w.defEncoder.Flush()
+		w.defLevelSink.SetOffset(0)
+		sz := w.defEncoder.Len()
+		if isV1DataPage {
+			sz += arrow.Uint32SizeBytes
+			binary.LittleEndian.PutUint32(w.defLevelSink.Bytes(), uint32(w.defEncoder.Len()))
+		}
+		defLevelsRLESize = int64(sz)
+	}
+
+	if w.descr.MaxRepetitionLevel() > 0 {
+		w.repEncoder.Flush()
+		w.repLevelSink.SetOffset(0)
+		if isV1DataPage {
+			binary.LittleEndian.PutUint32(w.repLevelSink.Bytes(), uint32(w.repEncoder.Len()))
+		}
+		repLevelsRLESize = int64(w.repLevelSink.Len())
+	}
+
+	uncompressed := defLevelsRLESize + repLevelsRLESize + int64(values.Len())
+	if isV1DataPage {
+		w.buildDataPageV1(defLevelsRLESize, repLevelsRLESize, uncompressed, values.Bytes())
+	} else {
+		w.buildDataPageV2(defLevelsRLESize, repLevelsRLESize, uncompressed, values.Bytes())
+	}
+
+	w.init()
+	w.numBuffered, w.numBufferedEncoded = 0, 0
+	return nil
+}
+
+func (w *columnWriter) buildDataPageV1(defLevelsRLESize, repLevelsRLESize, uncompressed int64, values []byte) error {
+	w.uncompressedData.Reset()
+	w.uncompressedData.Grow(int(uncompressed))
+	w.concatBuffers(defLevelsRLESize, repLevelsRLESize, values, &w.uncompressedData)
+
+	pageStats, err := w.getPageStatistics()
+	if err != nil {
+		return err
+	}
+	pageStats.ApplyStatSizeLimits(int(w.props.MaxStatsSizeFor(w.descr.Path())))
+	pageStats.Signed = schema.SortSIGNED == w.descr.SortOrder()
+	w.resetPageStatistics()
+
+	var data []byte
+	if w.pager.HasCompressor() {
+		w.compressedTemp.Reset()
+		data = w.pager.Compress(w.compressedTemp, w.uncompressedData.Bytes())
+	} else {
+		data = w.uncompressedData.Bytes()
+	}
+
+	// write the page to sink eagerly if there's no dictionary or if dictionary encoding has fallen back
+	if w.hasDict && !w.fallback {
+		pageSlice := make([]byte, len(data))
+		copy(pageSlice, data)
+		page := NewDataPageV1WithStats(memory.NewBufferBytes(pageSlice), int32(w.numBuffered), w.encoding, parquet.Encodings.RLE, parquet.Encodings.RLE, uncompressed, pageStats)
+		w.totalCompressedBytes += int64(page.buf.Len()) // + size of Pageheader
+		w.pages = append(w.pages, page)
+	} else {
+		w.totalCompressedBytes += int64(len(data))
+		dp := NewDataPageV1WithStats(memory.NewBufferBytes(data), int32(w.numBuffered), w.encoding, parquet.Encodings.RLE, parquet.Encodings.RLE, uncompressed, pageStats)
+		defer dp.Release()
+		w.WriteDataPage(dp)
+	}
+	return nil
+}
+
+func (w *columnWriter) buildDataPageV2(defLevelsRLESize, repLevelsRLESize, uncompressed int64, values []byte) error {
+	var data []byte
+	if w.pager.HasCompressor() {
+		w.compressedTemp.Reset()
+		data = w.pager.Compress(w.compressedTemp, values)
+		// data = w.compressedTemp.Bytes()
+	} else {
+		data = values
+	}
+
+	// concatenate uncompressed levels and the possibly compressed values
+	combined := make([]byte, int(defLevelsRLESize+repLevelsRLESize+int64(len(data))))
```

Comment:
is there a reason not to use concatBuffers here?

Reply:
modified to use `concatBuffers`